docs/proposals/memory-management.md

```diff
@@ -0,0 +1,185 @@
+
+   a. `Guaranteed`: Memory is always guaranteed up to the request. These containers will not OOM unless they exceed their request, or system processes consume too much memory, or the kernel overhead is high and the memory in the node is being fully consumed. Note that `request` equals `limit` for this class of containers.
+
+   b. `Burstable`: Memory is guaranteed up until the usage exceeds user specified `request`. These containers do not get killed as soon as they exceed their request. Kernel memory overhead and system processes memory usage will impact memory availability for this class of containers. They are more likely to be killed by the kernel OOM killer once they exceed their request.
+
+   c. `Best-Effort`: Memory is never guaranteed. This class of containers are the first ones to be killed by the kernel whenever there is memory pressure on the node.
+
+6. Detect OOM kills on containers: Docker tracks OOM kills on its containers.
+Kubelet exposes this information via [ContainerStatus.State](../../pkg/api/v1/types.go#L1033), where the state is set to `ContainerStateTerminated` and `Reason` is set to `OOM Killed`.
+Kubelet has a built-in exponential backoff mechanism that prevents OOMing containers from entering a crash loop and bringing down the node.
+
+
+#### What are some of the known limitations?
+
+1. The scheduler does not take into account the total usage on the node. Hence, memory can be easily over-committed. Whenever a job is scheduled without resource requests, the node becomes over-committed.
+
+2. Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+
+3. A container in a pod with restart policy set to `Always` will be restarted continuously even if it OOMs. This is alleviated with the crash loop backoff feature. To improve node stability, that pod should be evicted and run on a different node.
+
+4. Other than the Guaranteed class, containers in other classes don't have a memory limit.
+A `Best-effort` container can use a lot of memory and result in high tail latencies for `Guaranteed` and `Burstable` class containers.
+Memory allocations will hang until the kernel can free up memory, by identifying and killing a process in a container in a lower class.
+
+5. Kubelet does not act on OOM kills.
+A container can choose to continue running when one or more processes in the container were killed by the kernel OOM killer.
+Because of this, the kubelet cannot deterministically associate container death with OOMs.
+This affects higher level systems like auto-scaler which requires reliable OOM signals to update memory limits.
+
+6. In the event of an OOM kill due to system memory pressure, the kubelet cannot identify the container of the killed process.
+
+7. All burstable containers that exceed their request are equally likely to be killed. Ideally, the job that exceeds the most should be killed.
```

Comment:
I think this relates to https://github.com/kubernetes/kubernetes/pull/13006/files#r41343269

We should remove the "they are more likely to be killed when they exceed their specified request" from the above section.  Or rephrase, "they are more likely to be killed if any container in the same class exceeds its specified request"